### PR TITLE
FastBaumWelch loss as PyTorch autograd function

### PR DESCRIPTION
**Pull request by DanEnergetics**
Add a FastBaumWelch loss as a `torch.autograd` function and a simple wrapper for it. This can be imported and used in training directly.
Additionally add autoload mechanism.
Add test case.

### CODE REVIEW

File: i6_native_ops/fbw/__init__.py [hunk 1]
```diff
@@ -1,2 +1,65 @@
...
+
+def fbw_loss(
+    neg_log_probs: torch.FloatTensor,
+    fsa: Tuple[int, torch.IntTensor, torch.FloatTensor, torch.IntTensor],
+    seq_lens: torch.IntTensor
+) -> Tuple[torch.IntTensor, torch.FloatTensor]:
+    """ 
+    Computes negative log likelihood of an emission model given an HMM finite state automaton.
+    The corresponding gradient with respect to the emission model is automatically backpropagated.
+    :param log_probs: negative log probabilities of emission model as a (T, B, F)
```
Comment (michelwi) [inline i6_native_ops/fbw/__init__.py]:
```suggestion
    :param neg_log_probs: negative log probabilities of emission model as a (T, B, F)
```
to be consistent with actual parameter name.

However, would it be more consistent to other ops/losses in PT if instead the (positive) log_probs were being used? We could then in this function multiply with -1 if negative scores were required.

File: i6_native_ops/fbw/__init__.py [hunk 2]
```diff
@@ -1,2 +1,65 @@
...
+            os.path.join(base_path, "fbw_op.cu"),
+        ],
+        extra_include_paths=[
+            base_path,
+            os.path.join(base_path, "..", "common")
+        ],
+    )
+
+
+class FastBaumWelchLoss(torch.autograd.Function):
```
Comment (michelwi) [inline i6_native_ops/fbw/__init__.py:27]: I am no expert on custom PT ops, but should the fbw implementation not be similar to the fast_viterbi implementation in  #3?

File: i6_native_ops/fbw/__init__.py (hunk 1 above)
Comment (DanEnergetics) [reply i6_native_ops/fbw/__init__.py]: Yes I agree that accepting positive log probabilities would be less confusing here.

File: i6_native_ops/fbw/__init__.py (hunk 2 above)
Comment (DanEnergetics) [reply i6_native_ops/fbw/__init__.py:27]: The extra thing for fbw is that it backpropagates the gradients automatically. We can just register the output of `fbw_loss` as a loss and RETURNN/PyTorch handles the rest. The fast viterbi procedure does not give us a differentiable function and we're not interested in the gradients, that's why we can just wrap it into a simple function, giving us alignments and scores.

Comment (michelwi) [reply i6_native_ops/fbw/__init__.py:27]: >  viterbi [..] not [..] differentiable

yep makes sense 👍

**Review by michelwi — APPROVED** [review]

**Comment by michelwi** [comment]
Anybody else wants to review or can we merge as is?

File: i6_native_ops/fbw/__init__.py [hunk 3]
```diff
@@ -1,2 +1,68 @@
...
+        # negative log prob -> prob
+        grad = ctx.saved_tensors[0].neg().exp()
+        return grad, None, None
+
+
+def fbw_loss(
+    log_probs: torch.FloatTensor,
+    fsa: Tuple[int, torch.IntTensor, torch.FloatTensor, torch.IntTensor],
+    seq_lens: torch.IntTensor
+) -> Tuple[torch.IntTensor, torch.FloatTensor]:
```
Comment (JackTemaki) [inline i6_native_ops/fbw/__init__.py]: The return type seems wrong to me, is this not only one element?

Comment (DanEnergetics) [reply i6_native_ops/fbw/__init__.py]: Yes you're right

**Review by JackTemaki — APPROVED** [review]
